File: crypto/src/main/java/io/oversec/one/crypto/gpg/GpgEncryptionParams.java
package io.oversec.one.crypto.gpg;

import android.content.Context;
import io.oversec.one.crypto.AbstractEncryptionParams;
import io.oversec.one.crypto.EncryptionMethod;

import java.util.*;

public class GpgEncryptionParams extends AbstractEncryptionParams {


    long mOwnPublicKey;
    boolean mSign;
    Set<Long> mPublicKeys = new HashSet<>();

    public GpgEncryptionParams(List<Long> pkids, String coderId, String padderId) {
        super(EncryptionMethod.GPG, coderId, padderId);
        mPublicKeys.addAll(pkids);
    }

    public GpgEncryptionParams(long[] keyIds, String coderId, String padderId) {
        super(EncryptionMethod.GPG, coderId, padderId);
        for (long keyId : keyIds) {
            mPublicKeys.add(keyId);
        }
    }


    public Set<Long> getPublicKeyIds() {
        return mPublicKeys;
    }

    public void removePublicKey(Long keyId) {
        mPublicKeys.remove(keyId);
    }

    public Long getOwnPublicKey() {
        return mOwnPublicKey;
    }

    public void addPublicKeyIds(Long[] keyIds, long omitThisKey) {

        for (Long keyId : keyIds) {
            if (!keyId.equals(omitThisKey)) {
                mPublicKeys.add(keyId);
            }
        }
    }

    public static Long[] LongListToLongArray(List<Long> a) {
        if (a == null) {
            return null;
        }
        Long[] res = new Long[a.size()];
        a.toArray(res);
        return res;
    }

    public static List<Long> LongArrayToLongList(Long[] a) {
        if (a == null) {
            return null;
        }
        return new ArrayList<>(Arrays.asList(a));
    }

    public static Long[] longArrayToLongArray(long[] a) {
        if (a == null) {
            return null;
        }
        Long[] res = new Long[a.length];
        for (int i = 0; i < res.length; i++) {
            res[i] = a[i];
        }
        return res;
    }

    public static long[] LongArrayTolongArray(Long[] a) {
        if (a == null) {
            return null;
        }
        long[] res = new long[a.length];
        for (int i = 0; i < res.length; i++) {
            res[i] = a[i];
        }
        return res;
    }

    public void setOwnPublicKey(long k) {
        mOwnPublicKey = k;
        mPublicKeys.remove(k);
    }

    public void setSign(boolean s) {
        mSign = s;
    }

    public long[] getAllPublicKeyIds() {
        Set<Long> allKeys = new HashSet<>(mPublicKeys);

        if (mOwnPublicKey != 0) {
            allKeys.add(mOwnPublicKey);
        }
        long[] res = new long[allKeys.size()];
        int i = 0;
        for (long id : allKeys
                ) {
            res[i] = id;
            i++;
        }
        return res;


    }

    public boolean isSign() {
        return mSign;
    }

    public void addPublicKey(Long id) {
        mPublicKeys.add(id);
    }

    @Override
    public boolean isStillValid(Context ctx) {
        //TODO: check that OpenKEychain service is still up and running,
        //TODO: check that we do still have the keys
        return true;
    }

    public void setPublicKeyIds(long[] ids) {
        mPublicKeys.clear();
        for (long id : ids) {
            mPublicKeys.add(id);
        }
    }

    @Override
    public String toString() {

        return "GpgEncryptionParams{" +
                "mOwnPublicKey=" + mOwnPublicKey +
                ", mSign=" + mSign +
                ", mPublicKeys=" + Arrays.toString(mPublicKeys.toArray()) +
                '}';
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/gpg/GpgEncryptionParams.kt
package io.oversec.one.crypto.gpg

import android.content.Context
import io.oversec.one.crypto.AbstractEncryptionParams
import io.oversec.one.crypto.EncryptionMethod
import java.util.*

class GpgEncryptionParams : AbstractEncryptionParams {

    private var mOwnPublicKey: Long = 0
    var isSign = true
    private var mPublicKeys: MutableSet<Long> = HashSet()


    val publicKeyIds: Set<Long>
        get() = mPublicKeys

    val ownPublicKey: Long?
        get() = mOwnPublicKey

    val allPublicKeyIds: LongArray
        get() {
            val allKeys = ArrayList(mPublicKeys)
            if (mOwnPublicKey != 0L) {
                allKeys.add(mOwnPublicKey)
            }
            return allKeys.toLongArray()
        }

    constructor(pkids: List<Long>?, coderId: String, padderId: String?) : super(
        EncryptionMethod.GPG,
        coderId,
        padderId
    ) {
        pkids?.run { mPublicKeys.addAll(this) }

    }

    constructor(keyIds: LongArray?, coderId: String, padderId: String?) : super(
        EncryptionMethod.GPG,
        coderId,
        padderId
    ) {
        keyIds?.run {
            mPublicKeys.addAll(this.asList())
        }

    }

    fun removePublicKey(keyId: Long?) {
        mPublicKeys.remove(keyId)
    }

    fun addPublicKeyIds(keyIds: Array<Long>, omitThisKey: Long) {
        keyIds.forEach {
            if (it != omitThisKey) {
                mPublicKeys.add(it)
            }
        }
    }

    fun setOwnPublicKey(k: Long) {
        mOwnPublicKey = k
        mPublicKeys.remove(k)
    }

    fun addPublicKey(id: Long) {
        mPublicKeys.add(id)
    }

    override fun isStillValid(ctx: Context): Boolean {
        //TODO: check that OpenKEychain service is still up and running,
        //TODO: check that we do still have the keys
        return true
    }

    fun setPublicKeyIds(ids: LongArray) {
        mPublicKeys.clear()
        mPublicKeys.addAll(ids.asList())
    }

    override fun toString(): String {

        return "GpgEncryptionParams{" +
                "mOwnPublicKey=" + mOwnPublicKey +
                ", mSign=" + isSign +
                ", mPublicKeys=" + Arrays.toString(mPublicKeys.toTypedArray()) +
                '}'.toString()
    }

    companion object {

        fun LongListToLongArray(a: List<Long>?): LongArray? {
            return a?.toLongArray()
        }

        fun LongArrayToLongList(a: LongArray?): List<Long>? {
            return a?.toList()
        }

//        fun longArrayToLongArray(a: LongArray?): Array<Long>? {
//            if (a == null) {
//                return null
//            }
//            val res = arrayOfNulls<Long>(a.size)
//            for (i in res.indices) {
//                res[i] = a[i]
//            }
//            return res
//        }
//
//        fun LongArrayTolongArray(a: Array<Long>?): LongArray? {
//            if (a == null) {
//                return null
//            }
//            val res = LongArray(a.size)
//            for (i in res.indices) {
//                res[i] = a[i]
//            }
//            return res
//        }
    }
}
